src/error.rs

```diff
@@ -23,12 +23,12 @@ pub type Result<T> = std::result::Result<T, Error>;
 /// An error that can occur in the `mongodb` crate. The inner
 /// [`ErrorKind`](enum.ErrorKind.html) is wrapped in an `Arc` to allow the errors to be
 /// cloned.
-#[derive(Clone, Debug, Error)]
```

Comment:
Before we go through with removing the `Clone` implementation altogether, have we revisited just wrapping the `std::io::Error` branch in an `Arc` to preserve `Clone` on our type? I went back to see [why we didn't opt for this originally](https://docs.google.com/document/d/1SexV0QyBbb8BlhzDD9Wfib7eL3b8Hc_ltoU98wXGqtk/edit#heading=h.domcvrirm8h0), and it seems to be to avoid having users handle the indirection of `Arc` non-uniformally. For `std::io::Error` specifically, I'm not sure if `Arc` actually introduces any additional work for users though, since they have to invoke a method on the error (e.g. `kind()`) to get any useful result anyways.

I've looked over the other branches of our `ErrorKind`, and the only ones that also don't implement `Clone` are our BSON errors (which we own and can make `Clone`) and `TokioTimeoutElapsed` error, which we don't actually even use since `RUNTIME.timeout` converts everything to `std::io::Error` anyways.

I think if we manage to keep `Clone` alongside the changes made here to ease in matching, we'll have the best of both worlds.

Reply:
I considered that a bit too when working on this PR. I didn't end up making the change (although I hadn't looked back into why we didn't choose that originally), although a bit of that did make it into one of the proposals I made in the google doc in the context of making `CommandFailedEvent` use an alternate error struct.

I guess my main question is what benefit we think users will get from having our error type implement `Clone`. The only thing I can think of is the reason that we wanted it in the first place (to make their own wrapping error type cloneable), but I don't think this would be quite as common an issue for our users due to the fact that most of them are using the driver in the context of an application rather than a library, and the current consensus in the Rust community is that applications should just use `Box<dyn Error>` (or something like [`anyhow`](https://crates.io/crates/anyhow) to abstract that). While there are some libraries out there that wrap the driver (and obviously more could be written in the future), my general instinct is to make any performance costs (e.g. heap allocations and reference counts) opt-in for users who need them rather than on by default with no way to opt out of, especially since the number of users who will need it in this case are far fewer than the ones who won't. My opposition to doing this isn't particularly strong, though.

Reply:
I can't really think of a specific example of what a user might need the cloning for, but I imagine it could be something similar to what we needed clone for as you mentioned. Even if they don't _require_ the cloning (as we didn't), having to update their existing code could be non-trivial, as is seen in this PR. Also, if they really do need cloning, they'll need to wrap our entire error in an `Arc`, which is pretty annoying to deal with (hence this PR :slightly_smiling_face:), so even if it is somewhat uncommon, the frustration felt could be high and worth avoiding. Lastly, if we don't go for `Clone` now, I don't think we'll be able to do so until 3.0, so if it does turn out to be a burden for users, we can't really give them any good fix.

Regarding performance costs, I definitely agree, but given that this Arc would only be allocated in single specific error case (i.e. not the "hot" path), I don't think it's much of a concern.

So overall, for reasons of reducing user friction + safeguarding us against any future need, I think we should try to preserve `Clone` here.

This also has made me wonder, if we preserve `Clone`, could we avoid having to make the `String` changes in this PR? I realize this is an frustrating question to ask _after_ all the work has been done and I'm really sorry about that; it's just that up to this point I had assumed that preserving `Clone` wasn't going to be possible or easy but now have realized there could be another option.

Reply:
If we kept Error being Clone, then yes, I think we would want to undo the changes to CommandFailedEvent. All of the other changes are probably fine to keep though, since they don't put any burden on the user and are marginal perf improvements.